openpype/hosts/maya/plugins/publish/extract_workfile_xgen.py

```diff
@@ -0,0 +1,250 @@
+
+            if start_frame is None:
+                start_frame = render_start_frame
+            else:
+                start_frame = min(start_frame, render_start_frame)
+
+            if end_frame is None:
+                end_frame = render_end_frame
+            else:
+                end_frame = max(end_frame, render_end_frame)
+
+        if start_frame is None or end_frame is None:
+            return
+
+        return start_frame, end_frame
+
+    def process(self, instance):
+        transfers = []
+
+        # Validate there is any palettes in the scene.
+        if not cmds.ls(type="xgmPalette"):
+            self.log.debug(
+                "No collections found in the scene. Skipping Xgen extraction."
+            )
+            return
+        else:
+            import xgenm
+
+        # Validate to extract only when we are publishing a renderlayer as
+        # well.
+        render_range = self.get_render_max_frame_range(instance.context)
+        if not render_range:
+            self.log.debug(
+                "No publishable renderlayers found in context. Skipping Xgen"
+                " extraction."
+            )
+            return
+
+        start_frame, end_frame = render_range
+
+        # We decrement start frame and increment end frame so motion blur will
+        # render correctly.
+        start_frame -= 1
+        end_frame += 1
+
+        # Extract patches alembic.
+        basename, _ = os.path.splitext(instance.context.data["currentFile"])
+        dirname = os.path.dirname(instance.context.data["currentFile"])
+        kwargs = {"attrPrefix": ["xgen"], "stripNamespaces": True}
+        alembic_files = []
+        for palette in cmds.ls(type="xgmPalette"):
+            patch_names = []
+            for description in xgenm.descriptions(palette):
+                for name in xgenm.boundGeometry(palette, description):
+                    patch_names.append(name)
+
+            alembic_file = os.path.join(
+                dirname,
+                "{}__{}.abc".format(basename, palette.replace(":", "__ns__"))
+            )
+            extract_alembic(
+                alembic_file,
+                root=patch_names,
+                selection=False,
+                startFrame=float(start_frame),
+                endFrame=float(end_frame),
+                verbose=True,
+                **kwargs
+            )
+            alembic_files.append(alembic_file)
+
+        template_data = copy.deepcopy(instance.data["anatomyData"])
+        published_maya_path = StringTemplate(
+            instance.context.data["anatomy"].templates["publish"]["file"]
+        ).format(template_data)
+        published_basename, _ = os.path.splitext(published_maya_path)
+
+        for source in alembic_files:
+            destination = os.path.join(
+                os.path.dirname(instance.data["resourcesDir"]),
+                os.path.basename(source.replace(basename, published_basename))
```

Comment:
Reading this made me think "so wait, `basename` and `publish_basename` themselves aren't actually basenames"? They didn't go through `os.path.basename` but are the `root` of `os.path.splitext`. Do we have naming confusion here?

See this example of `os.path.splitext`:
```python
splitext('/foo/bar.exe')
# ('/foo/bar', '.exe')
```

Reply:
Have you got a suggestions of names/code?